### PR TITLE
Implement NoFail, HardRock, Easy mods

### PR DESCRIPTION
**Pull request by peppy**
Restructures beatmap handling a bit, ensures local changes remain local via a DeepClone of BeatmapInfo inside BeatmapConverter.

### CODE REVIEW

File: osu.Game/Rulesets/Mods/Mod.cs [hunk 1]
```diff
@@ -45,5 +45,10 @@ public abstract class Mod
         /// The mods this mod cannot be enabled with.
         /// </summary>
         public virtual Type[] IncompatibleMods => new Type[] { };
+
+        /// <summary>
+        /// Whether we should allow fails at the
```
Comment (MrTheMake) [inline osu.Game/Rulesets/Mods/Mod.cs]: Something is missing here.

File: osu.Game/Graphics/SpriteIcon.cs [hunk 2]
```diff
@@ -60,10 +62,32 @@ private void updateTexture()
...
-            set { spriteShadow.Alpha = value ? 1 : 0; }
+            set
+            {
+                if (value == (spriteShadow.IsAlive && spriteShadow.IsLoaded)) return;
+
+                if (value)
+                    AddInternal(spriteShadow);
+                else
+                    RemoveInternal(spriteShadow);
+            }
```
Comment (smoogipooo) [inline osu.Game/Graphics/SpriteIcon.cs]: Is this stuff really needed? I don't see the point of not adding spriteShadow as it was previously being done.

Comment (peppy) [inline osu.Game/Graphics/SpriteIcon.cs]: it's basically the same thing right? just saves on storing alpha to a local

File: osu.Game/Graphics/SpriteIcon.cs [hunk 3]
```diff
@@ -60,10 +62,32 @@ private void updateTexture()
...
+            return base.Invalidate(invalidation, source, shallPropagate);
+        }
+
         public bool Shadow
         {
             get { return spriteShadow.IsPresent; }
-            set { spriteShadow.Alpha = value ? 1 : 0; }
+            set
+            {
+                if (value == (spriteShadow.IsAlive && spriteShadow.IsLoaded)) return;
```
Comment (smoogipooo) [inline osu.Game/Graphics/SpriteIcon.cs]: This is not non-async-load-safe. I can break this with the following code in a testcase ctor (or really, any ctor):
```
SpriteIcon sr;
Add(sr = new SpriteIcon
{
    Icon = FontAwesome.fa_question,
    Size = new Vector2(100)
});

for (int i = 0; i < 1000; i++)
    sr.Shadow = true;
```

Comment (smoogipooo) [inline osu.Game/Graphics/SpriteIcon.cs]: I think in the end you're going to have to store a local in some way, so might as well rework this method to not do the whole adding + removal as well.

Comment (smoogipooo) [inline osu.Game/Graphics/SpriteIcon.cs]: Consider wrapping the shadow in another container, and setting the alpha of that thing. ¯\_(ツ)_/¯

**Review by smoogipoo — CHANGES_REQUESTED** [review]
Have considered and am fine with the rest of the PR, just need to fix that SpriteIcon shadow property and good to go!

**Review by smoogipoo — APPROVED** [review]
lgtm

File: osu.Game/Rulesets/Mods/ModHardRock.cs [hunk 4]
```diff
@@ -2,16 +2,26 @@
...
         public override string Name => "Hard Rock";
         public override FontAwesome Icon => FontAwesome.fa_osu_mod_hardrock;
         public override ModType Type => ModType.DifficultyIncrease;
         public override string Description => "Everything just got a bit harder...";
         public override Type[] IncompatibleMods => new[] { typeof(ModEasy) };
+
+        public void ApplyToDifficulty(BeatmapDifficulty difficulty)
+        {
+            const float ratio = 1.4f;
+            difficulty.CircleSize *= ratio;
```
Comment (MillhioreF) [inline osu.Game/Rulesets/Mods/ModHardRock.cs]: CircleSize is only multiplied by 1.3f in stable. Guessing this is an oversight?

Comment (smoogipooo) [inline osu.Game/Rulesets/Mods/ModHardRock.cs]: Today I learnt O_O

Comment (peppy) [inline osu.Game/Rulesets/Mods/ModHardRock.cs]: Ah, I saw this but thought it was specific to the cursor size adjustment. Should probably be changes, yes.

**Pull request merged**